public/ledger.js

```diff
@@ -0,0 +1,290 @@
+const constants = require('./constants');
+
+const maxLedgerBuffer = 255;
+
+class Ledger {
+  
+  static parseLedgerError(e) {
+    if (e.name && e.name === 'TransportStatusError') {
+      switch (e.statusCode) {
+        case 0x6985:
+          return {status: e.statusCode, message: 'Request denied by user on Ledger'};
+        case 0x6b00:  // SW_DEVELOPER_ERROR
+        case 0x6b01:  // SW_INVALID_PARAM
+        case 0x6b02:  // SW_IMPROPER_INIT
+        default:
+          return {status: e.statusCode, message: 'Error communicating with Ledger'};
+      }
+    }
+    return e;
+  }
+
+  constructor() {
+    // Electron main window to send command to renderer process
+    this.mainWindow = null;
+    // Ledger transport
+    this.transport = null;
+    // Timeout for ledger connection
+    this.timeout = 10000;
+    // If hathor app is opened on ledger
+    this.hathorAppOpened = false;
+
+    this.subscriptor = Transport.listen({
+      next: e => {
+        if (e.type === 'add') {
+          // All events called here we check if we are in the hathor app
+          this.getVersion();
+        } else if (e.type === 'remove') {
+          // This is called in some situations
+          // - Usb disconnected
+          // - When user quits an app
+          // - When user opens an app
+          if (this.hathorAppOpened) {
+            // If was opened we schedule a setTimeout
+            // to lock the wallet in case it continues closed
+            setTimeout(() => {
+              if (!this.hathorAppOpened) {
+                this.mainWindow.webContents.send("ledger:closed");
+              }
+            }, 3000);
+          }
+          this.hathorAppOpened = false;
+          // If we don't close it when changing the app we get an error
+          // https://github.com/LedgerHQ/ledgerjs/issues/22
+          this.transport.device.close();
+          this.transport = null;
+        }
+      },
+      error: error => {},
+      complete: () => {}
+    })
+
+    // Ledger commands
+    this.commands = {
+      'VERSION': 0x01,
+      'ADDRESS': 0x02,
+      'SEND_TX': 0x04,
+      'PUBLIC_KEY_DATA': 0x10,
+    }
+
+    // Queue of commands to send to ledger so we don't send them in paralel
+    this.sendQueue = [];
+
+    // If can send command to ledger
+    this.canSend = true;
+  }
+
+  checkSendQueue = () => {
+    if (this.sendQueue.length > 0) {
+      const { transport, command, p1, p2, data, resolve, reject } = this.sendQueue.shift();
+      const promise = transport.send(0xe0, command, p1, p2, data);
+      promise.then((response) => {
+        resolve(response);
+      }, (error) => {
+        reject(error);
+      }).finally(() => {
+        this.checkSendQueue();
+      });
+    } else {
+      this.canSend = true;
+    }
+  }
+
+  sendToLedgerOrQueue = (transport, command, p1, p2, data) => {
+    if (this.canSend) {
+      this.canSend = false;
+      const promise = transport.send(0xe0, command, p1, p2, data);
```

Comment:
What is `0xe0`? Shouldn't we create a constant?

Reply:
Done